### PR TITLE
RedisConnection: set TCP_NODELAY

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
Seems to give a little boost.

File: lib/icingadb/redisconnection.cpp [hunk 1]
```diff
@@ -226,6 +226,10 @@ void RedisConnection::Connect(asio::yield_context& yc)
 					<< "Trying to connect to Redis server (async) on host '" << m_Host << ":" << m_Port << "'";
 
 				auto conn (Shared<TcpConn>::Make(m_Strand.context()));
+
+				const int optTrue = 1;
+				setsockopt(conn->lowest_layer().native_handle(), IPPROTO_TCP, TCP_NODELAY, reinterpret_cast<const char *>(&optTrue), sizeof(optTrue));
```
Comment (julianbrost) [inline lib/icingadb/redisconnection.cpp:231]: Why do you explicitly cast there? `setsockopt()` takes a `const void *` so just `&optTrue` should be perfectly fine here.

**Comment by julianbrost** [comment]
@Al2Klimov Can you please quantify "a little boost" and did you test this with this branch as is or with others merged? I suggested to try `TCP_NODELAY` in case we'd use `HSCAN` where you care for the latency of the result. But in this branch, there seems to be no use of `RedisConnection::{GetResultOfQuery,GetResultsOfQueries}`, so I'd expect little to no difference.

**Comment by Al2Klimov** [comment]
<1%, I tested vs. the parent commit. Not in this branch, but in #8650.

**Comment by julianbrost** [comment]
If you want to do me a favor, you can measure the time of each HSCAN call, sum these values over the whole loop and share the numbers with and without TCP_NODELAY (otherwise I'll do this tomorrow): https://github.com/Icinga/icinga2/blob/d022e34a906c14549c539b823730826cc0c5df56/lib/icingadb/icingadb-objects.cpp#L175-L177

**Comment by julianbrost** [comment]
Just tried what I suggested and I came to the conclusion that currently there's way too much jitter in the times to draw a conclusion from just a few measurements.

```
[2021-03-04 11:57:56 +0000] information/RedisClient: HSCAN icinga:checksum:service took 3648.4ms in 11 iterations
[2021-03-04 11:57:57 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4105.86ms in 11 iterations
[2021-03-04 11:58:25 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4123.48ms in 11 iterations
[2021-03-04 11:58:25 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4134.23ms in 11 iterations
[2021-03-04 11:59:20 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4585.9ms in 11 iterations
[2021-03-04 11:59:20 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4743.14ms in 11 iterations
[2021-03-04 12:00:09 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4622.84ms in 11 iterations
[2021-03-04 12:00:09 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4833.5ms in 11 iterations
[2021-03-04 12:00:25 +0000] information/RedisClient: HSCAN icinga:checksum:service took 3926.45ms in 11 iterations
[2021-03-04 12:00:25 +0000] information/RedisClient: HSCAN icinga:checksum:service took 3939.87ms in 11 iterations
[2021-03-04 12:00:51 +0000] information/RedisClient: HSCAN icinga:checksum:service took 3892.46ms in 11 iterations
[2021-03-04 12:00:51 +0000] information/RedisClient: HSCAN icinga:checksum:service took 4127.5ms in 11 iterations
```

Then I noticed that #8638 is still a draft and my suggestion to try `TCP_NODELAY` only really makes sense if that's merged and there's a risk that the connection is idle after an HSCAN query is issued. But with the current single connections that's not the case anyways.

So I think the <1% difference you observed was just random jitter, and we should postpone this PR until we're in a situation where using TCP_NODELAY might have an effect and then reevaluate this.

**Comment by julianbrost** [comment]
I originally suggested trying `TCP_NODELAY` because the `HSCAN` on a single connection pattern could potentially introduce some 40ms latency on Linux (see for example [[1]](https://vorner.github.io/2020/11/06/40-ms-bug.html), [[2]](https://jvns.ca/blog/2015/11/21/why-you-should-understand-a-little-about-tcp/), [[3]](https://rachelbythebay.com/w/2020/10/14/lag/)).

I've added logging for the actual send query and read response cycle and ran it with and without this PR (merged onto the current master). There is no 40ms latency (almost all queries have sub-millisecond latency) nor do I see a significant difference. Therefore, I'm closing this PR.

#### Without TCP_NODELAY
```
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0181198ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:hostgroup' '0' 'COUNT' '1000' took 0.0350475ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.025034ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:downtime' '0' 'COUNT' '1000' took 0.0379086ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:eventcommand' '0' 'COUNT' '1000' took 0.0379086ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:zone' '0' 'COUNT' '1000' took 0.0348091ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0419617ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:servicegroup' '0' 'COUNT' '1000' took 0.0360012ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:user' '0' 'COUNT' '1000' took 0.0481606ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:notification' '0' 'COUNT' '1000' took 0.0319481ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.00500679ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:notificationcommand' '0' 'COUNT' '1000' took 0.041008ms 
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:checkcommand' '0' 'COUNT' '1000' took 3.36599ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.00500679ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:host' '0' 'COUNT' '1000' took 0.0269413ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.770092ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0138283ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:timeperiod' '0' 'COUNT' '1000' took 0.0488758ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:usergroup' '0' 'COUNT' '1000' took 0.041008ms 
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:endpoint' '0' 'COUNT' '1000' took 0.0560284ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.00405312ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0119209ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.00715256ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:service' '0' 'COUNT' '1000' took 0.12517ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0300407ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0200272ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.00500679ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.00405312ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0169277ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:comment' '0' 'COUNT' '1000' took 0.0331402ms
[2021-07-19 12:37:44 +0200] information/IcingaDB/Latency:  'PING' took 0.0369549ms
```

#### With TCP_NODELAY
```
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0119209ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:user' '0' 'COUNT' '1000' took 0.0350475ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:downtime' '0' 'COUNT' '1000' took 0.0400543ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:notificationcommand' '0' 'COUNT' '1000' took 0.038147ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0212193ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:hostgroup' '0' 'COUNT' '1000' took 0.0300407ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:checkcommand' '0' 'COUNT' '1000' took 0.0360012ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:eventcommand' '0' 'COUNT' '1000' took 0.0360012ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0350475ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.025034ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0321865ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0181198ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:servicegroup' '0' 'COUNT' '1000' took 0.0391006ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0150204ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:service' '0' 'COUNT' '1000' took 0.0720024ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:host' '0' 'COUNT' '1000' took 0.0472069ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:zone' '0' 'COUNT' '1000' took 0.0457764ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:notification' '0' 'COUNT' '1000' took 0.0700951ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:usergroup' '0' 'COUNT' '1000' took 0.0460148ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:endpoint' '0' 'COUNT' '1000' took 3.10898ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.0209808ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:timeperiod' '0' 'COUNT' '1000' took 0.0350475ms 
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.00882149ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.00596046ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.082016ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.00596046ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.00190735ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.00500679ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 0.00286102ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'HSCAN' 'icinga:checksum:comment' '0' 'COUNT' '1000' took 0.0548363ms
[2021-07-19 12:41:56 +0200] information/IcingaDB/Latency:  'PING' took 1.05ms
```

<details>
<summary>Patch to log query times</summary>

```diff
diff --git a/lib/icingadb/redisconnection.cpp b/lib/icingadb/redisconnection.cpp
index a605e7967..44b8d06e0 100644
--- a/lib/icingadb/redisconnection.cpp
+++ b/lib/icingadb/redisconnection.cpp
@@ -453,6 +453,14 @@ void RedisConnection::WriteItem(boost::asio::yield_context& yc, RedisConnection:
        if (next.GetResultOfQuery) {
                auto& item (*next.GetResultOfQuery);
 
+               double begin = Utility::GetTime();
+               Defer logTime([begin, &item]() {
+                       double end = Utility::GetTime();
+                       Log msg (LogInformation, "IcingaDB/Latency");
+                       LogQuery(item.first, msg);
+                       msg << " took " << (end - begin)*1000 << "ms";
+               });
+
                try {
                        WriteOne(item.first, yc);
                } catch (const boost::coroutines::detail::forced_unwind&) {
```
</details>

**Pull request closed without merging**